pygdf/queryutils.py

```diff
@@ -202,6 +203,15 @@ def query_execute(df, expr, callenv):
         name = name[len(ENVREF_PREFIX):]
         try:
             val = envdict[name]
+            if isinstance(val, dt.datetime):
+                # TODO right now we can't
+                # pass datetimes into numba so coerce to ms as int
+                # this is a todo because what if we want nanos?
+                val = int(time.mktime(val.timetuple()) * 1e3)
+            elif isinstance(val, np.datetime64):
+                from .datetime import DatetimeColumn
+                val = val.astype(DatetimeColumn._npdatetime64_dtype)
+                val = val.astype(DatetimeColumn._int_dtype)
```

Comment:
Given we have to cast to int64 for numba compatibility, should this behavior then throw an exception with a message telling the user to cast to int64 before running a query or UDF? Magically typecasting to int64 behind the curtains is going to lead to unexpected behavior for end users.

Reply:
yea i don't know - it's a pretty poor user experience to have to cast before you do any datetie ops, but I don't know what else can be done

Reply:
We should not make the user manually cast before querying against datetime columns.

One of the major things that makes DataFrames so popular is the ease of use that comes from implicit casting that happens behind the scenes.

I think we need to make the docs explicit that .query and UDFs called against DateTimes is performing a cast to int64.

Reply:
so what do we do?  I'm happy to do whatever y'all decide

Reply:
I'd also be in favor of doing a cat or convert as needed here,

Reply:
One issue with this type of casting is that i the gdf column is in a different precision than we choose to cast to, we're gonna do something terribly wrong.  In the code there isn't a good way to associate which objects are operating with which objects.  What we could do is either have the method that goes from gdf column to numba coerce the datetime column into some fixed precision (which matches how we would coerce datetimes).   Either that or we throw in that method if the data isn't in the proper precision.

Currently we call to_gpu_array to go from gdf column to something numba uses.  Would this be the right place to do this?   does to_gpu_array mean to a gpu array that numba will use in numba compiled exprs?  @sklam

Reply:
I believe @sklam is working on addressing this issue from within Numba. See: https://github.com/numba/numba/pull/3198

@sklam Would love for you to weigh in here to understand how your PR above could fit in with this.

Reply:
numba/numba#3198 fixed the problem with using np.datetime and np.timedelta dtypes inside numba cuda kernels.  That patch should mean that you don't need to cast to int64 anymore.

Reply:
@hhuuggoo I think instead of typecasting to a python datetime object you can typecast to a numpy datetime scalar and that should hopefully resolve these issues.